packages/@aws-cdk/aws-cloudfront/lib/origin.ts

```diff
@@ -205,3 +206,33 @@ function validateIntInRangeOrUndefined(name: string, min: number, max: number, v
     throw new Error(`${name}: Must be an int between ${min} and ${max}${seconds} (inclusive); received ${value}.`);
   }
 }
+
+/**
+ * Throws an error if custom header assignment is prohibited by CloudFront.
+ * @link: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/add-origin-custom-headers.html#add-origin-custom-headers-use-cases
+ */
+function validateCustomHeaders(customHeaders: Record<string, string>) {
+  const customHeaderKeys = Object.keys(customHeaders);
+  const prohibitedHeaderKeys = [
+    'Cache-Control', 'Connection', 'Content-Length', 'Cookie', 'Host', 'If-Match', 'If-Modified-Since', 'If-None-Match', 'If-Range', 'If-Unmodified-Since',
+    'Max-Forwards', 'Pragma', 'Proxy-Authorization', 'Proxy-Connection', 'Range', 'Request-Range', 'TE', 'Trailer', 'Transfer-Encoding', 'Upgrade', 'Via',
+    'X-Real-Ip',
+  ];
+  const prohibitedHeaderKeyPrefixes = [
+    'X-Amz-', 'X-Edge-',
+  ];
+
+  const prohibitedHeadersKeysMatches = customHeaderKeys.filter(customKey => {
+    return prohibitedHeaderKeys.map((prohibitedKey) => prohibitedKey.toLowerCase()).includes(customKey.toLowerCase());
+  });
+  const prohibitedHeaderPrefixMatches = customHeaderKeys.filter(customKey => {
+    return prohibitedHeaderKeyPrefixes.some(prohibitedKeyPrefix => customKey.toLowerCase().startsWith(prohibitedKeyPrefix.toLowerCase()));
+  });
+
+  if (prohibitedHeadersKeysMatches.length !== 0) {
+    throw new Error(`You can’t configure CloudFront to add any of the following headers to requests that it sends to your origin: ${prohibitedHeaderKeys.join(', ')}. Got: ${customHeaderKeys.join(', ')}.`);
```

Comment:
It seems it would be more useful to specifically mention which headers are the issue, rather than providing the complete lists of prohibited and provided headers and make the user do the intersection themselves.

```suggestion
    throw new Error(`You can’t configure CloudFront to add any of the following headers to requests that it sends to your origin: ${prohibitedHeaderKeys.join(', ')}. Got: ${prohibitedHeadersKeysMatches.join(', ')}.`);
```

I could even see shortening this down to:
```
    throw new Error(`the following headers cannot be configured as custom origin headers:  ${prohibitedHeadersKeysMatches.join(', ')}`);
```

Same idea for the next error below.